hi-ml-azure/src/health_azure/utils.py

```diff
@@ -895,54 +896,125 @@ def _log_conda_dependencies_stats(conda: CondaDependencies, message_prefix: str)
+            unique_dependencies.append(keep_dependency)
+    return unique_dependencies
+
+
+def _retrieve_unique_deps(dependencies: List[str], pinned_operator: PinnedOperator) -> List[str]:
     """
     Given a list of conda dependencies, which may contain duplicate versions
     of the same package name with the same or different versions, returns a
     list of them where each package name occurs only once. If a
-    package name appears more than once, only the first value will be retained.
+    package name appears more than once, a simple resolution strategy will be applied:
+    If any of the versions is listed with an equality constraint, that will be kept, irrespective
+    of the other constraints, even if they clash with the equality constraint. Multiple equality
+    constraints raise an error.
 
-    :param dependencies: The original list of package names to deduplicate
-    :param keep_method: The strategy for choosing which package version to keep
-    :return: a list in which each package name occurs only once
+    :param dependencies: the original list of package names to deduplicate
+    :return: a list of package specifications in which each package name occurs only once
     """
-    unique_deps: Dict[str, Tuple[str, str]] = {}
+    all_deps: DefaultDict[str, List[PackageDependency]] = defaultdict()
     for dep in dependencies:
-        dep_parts: List[str] = re.split("(=<|==|=|>=|<|>)", dep)
-        len_parts = len(dep_parts)
-        dep_name = dep_parts[0]
-        if len_parts > 1:
-            dep_join = "".join(dep_parts[1:-1])
-            dep_version = dep_parts[-1]
-        else:
-            dep_join = ""
-            dep_version = ""
-
-        if dep_name in unique_deps:
-            if keep_method == "first":
-                keep_version, _ = unique_deps[dep_name]
-            elif keep_method == "last":
-                keep_version = dep_version
-                unique_deps[dep_name] = (keep_version, dep_join)
-            else:
-                raise ValueError(
-                    f"Unrecognised value of 'keep_method: {keep_method}'. Accepted values"
-                    f" include: ['first', 'last']"
-                )
-            logging.warning(
-                f"Found duplicate requirements: {dep}. Keeping the {keep_method} " f"version: {keep_version}"
-            )
+        dependency = PackageDependency(dep)
 
+        dependency_name = dependency.package_name
+        if dependency_name in all_deps:
+            all_deps[dependency_name].append(dependency)
         else:
-            unique_deps[dep_name] = (dep_version, dep_join)
+            all_deps[dependency_name] = [dependency]
 
-    unique_deps_list = [f"{pkg}{joiner}{vrsn}" for pkg, (vrsn, joiner) in unique_deps.items()]
+    unique_deps: List[PackageDependency] = _resolve_dependencies(all_deps, pinned_operator)
+
+    unique_deps_list = [f"{dep.package_name}{dep.operator}{dep.version}" for dep in unique_deps]
```

Comment:
You could move that f"..." string into a __str__ method of the PackageDependency class